vaporetto/src/type_scorer/boundary_scorer_cache.rs

```diff
@@ -0,0 +1,112 @@
+use alloc::vec::Vec;
+
+use bincode::{Decode, Encode};
+use daachorse::DoubleArrayAhoCorasick;
+
+use crate::errors::{Result, VaporettoError};
+use crate::ngram_model::NgramModel;
+use crate::sentence::Sentence;
+
+const ALPHABET_SIZE: usize = 8;
+const ALPHABET_MASK: usize = ALPHABET_SIZE - 1;
+const ALPHABET_SHIFT: usize = 3;
+
+#[derive(Decode, Encode)]
+pub struct TypeScorerBoundaryCache {
+    scores: Vec<i32>,
+    window_size: u8,
+    sequence_mask: usize,
+}
+
+impl TypeScorerBoundaryCache {
+    pub fn new(model: NgramModel<Vec<u8>>, window_size: u8) -> Result<Self> {
+        let pma = DoubleArrayAhoCorasick::new(model.0.iter().map(|d| &d.ngram))
+            .map_err(|_| VaporettoError::invalid_model("invalid character type n-grams"))?;
+        let mut weights = vec![];
+        for d in model.0 {
+            if d.weights.len() <= 2 * usize::from(window_size) - d.ngram.len() {
```

Comment:
When `2 * usize::from(window_size) < d.ngram.len()`, the subtraction involves an underflow.
It would be necessary to guarantee that such underflow never occurs.

Reply:
I removed this conditional branch.
If `d.weights` is shorter than the expected length, 0 will be used instead.